signer/src/storage/memory/store.rs

```diff
@@ -0,0 +1,415 @@
+                .cloned()
+                .unwrap_or_default()
+                .into_iter()
+                .map(|pk| {
+                    self.withdrawal_requests
+                        .get(&pk)
+                        .expect("missing withdraw request")
+                        .clone()
+                })
+        })
+        .collect()
+    }
+}
+
+impl Transactable for SharedStore {
+    type Tx<'a> = InMemoryTransaction;
+
+    async fn begin_transaction(&self) -> Result<Self::Tx<'_>, Error> {
+        let store = self.lock().await;
+        let store_clone = store.clone();
+        Ok(InMemoryTransaction {
+            version: store.version,
+            store: Arc::new(Mutex::new(store_clone)),
+            original_store_mutex: Arc::clone(self),
+            completed: AtomicBool::new(false),
+        })
+    }
+}
+
+/// Represents an active in-memory transaction.
+pub struct InMemoryTransaction {
+    /// Records the version of the store at the time of transaction creation.
+    pub version: usize,
+    /// Holds a clone of the store's data for operations within this transaction.
+    pub store: SharedStore,
+    /// Reference to the original store's mutex to commit changes back.
+    pub original_store_mutex: SharedStore,
+    /// Tracks if commit/rollback has been called.
+    pub completed: AtomicBool,
+}
+
+impl TransactionHandle for InMemoryTransaction {
+    async fn commit(self) -> Result<(), Error> {
+        if self.completed.load(std::sync::atomic::Ordering::SeqCst) {
+            panic!("Transaction already completed");
+        }
+
+        // Lock the transaction's clone of the store and get a guard
+        let store = self.store.lock().await.clone();
+        // Lock the original store and get a guard
+        let mut original_store = self.original_store_mutex.lock().await;
+
+        // Naive optimistic concurrency check
+        if self.version != original_store.version {
+            return Err(Error::OptimisticConcurrencyViolation {
+                transaction_version: self.version,
+                store_version: original_store.version,
+            });
+        }
+
+        // Commit the changes from the transactional store to the original store.
+        // This copies all fields from transactional_store_content.
+        *original_store = store.clone();
+
+        // Explicitly set the original store's version to be one greater than the version
+        // this transaction started with. This marks that the store has been updated.
+        original_store.version = self.version + 1;
+
+        // Mark the transaction as completed.
+        self.completed
+            .store(true, std::sync::atomic::Ordering::SeqCst);
+
+        Ok(())
+    }
+
+    async fn rollback(self) -> Result<(), Error> {
+        if self.completed.load(std::sync::atomic::Ordering::SeqCst) {
+            panic!("Transaction already completed");
+        }
+
+        // Rollback is a no-op for in-memory transactions.
```

Comment:
Don't we still need to acquire a lock (maybe on `self.store`)? Otherwise we could concurrently execute `commit` and `rollback`; it should have no real consequence (as we don't do anything in here, and at the end in `commit` we would still mark completed as true), but it feels wrong.

Reply:
[d22566f](https://github.com/stacks-sbtc/sbtc/pull/1674/commits/d22566feef969c3b28fe36b7f19e74833b73be08)